File: hashTableFree.cpp
#include <iostream>
#include <algorithm>
#include <sstream>

using namespace std;

size_t genHashFree(int size, string key){
    size_t result = 1245;
    for (size_t i = 0; i < key.size(); i++){
        result += i*key[i] % size;
    }
    return result % size;
}

struct HashTableFree{
    pair<string, string>* table;
    size_t size;
    

    HashTableFree(size_t sz) : size(sz){
        table = new pair<string, string>[sz];
        for(int i = 0; i < size; i++) table[i].first = "_empty_";
    }
    ~HashTableFree(){
        delete[] table;
    }


    void insert(string key, string value){
        size_t Hash = genHashFree(size, key);
        if (table[Hash].first == key){
            table[Hash] = {key, value};
            cout << "Элемент добавлен: ключ='" << key << "', значение='" << value << "'" << endl;
            return;
        }
        if (table[Hash].first == "_empty_"){
            table[Hash] = {key, value};
            cout << "Элемент добавлен: ключ='" << key << "', значение='" << value << "'" << endl;
        }else{
            int i = 1;
            while(1){
                if (table[(Hash + i)%size].first == key){
                    table[Hash] = {key, value};
                    cout << "Элемент добавлен: ключ='" << key << "', значение='" << value << "'" << endl;
                    return;
                }
                if (table[(Hash + i)%size].first == "_empty_"){
                    table[(Hash + i)%size] = {key, value};
                    cout << "Элемент добавлен: ключ='" << key << "', значение='" << value << "'" << endl;
                    return;
                } 
                i++;
                if (i > size*2){
                    cout << "Свободное место отсутствует в таблице" << endl;
                    return;
                }
            }
        }
    }

    void remove(string key){
        size_t Hash = genHashFree(size, key);
        if (table[Hash].first == key){
            table[Hash].first = "_empty_";
            cout << "Элемент успешно удалён" << endl;
        }else{
            int i = 1;
            while(1){
                if (table[(Hash + i)%size].first == key){
                    table[(Hash + i)%size].first = "_empty_";
                    cout << "Элемент успешно удалён" << endl;
                    return;
                } 
                i++;
                if (i > size*2){
                    cout << "Элемент отсутствует в таблице" << endl;
                    return;
                }
            }
        }
    }

    void find(string key){
        size_t Hash = genHashFree(size, key);
        if (table[Hash].first == key){
            cout << "Data = " << table[Hash].second << endl;
        }else{
            int i = 1;
            while(1){
                if (table[(Hash + i)%size].first == key){
                    cout << "Data = " << table[(Hash + i)%size].second << endl;
                    return;
                } 
                i++;
                if (i > size*2){
                    cout << "Элемент отсутствует в таблице" << endl;
                    return;
                }
            }
        }
    }
};

void hashTableFree(){
    cout << "Введите размер хеш-таблицы:" << endl;
    
    size_t tableSize;
    cin >> tableSize;
    
    if (tableSize <= 0) {
        cout << "Ошибка: размер таблицы должен быть положительным числом" << endl;
        return;
    }
    
    HashTableFree hashTable(tableSize);
    cout << "Хеш-таблица создана с размером " << tableSize << endl;
    
    cin.ignore(); // Очищаем буфер после ввода числа
    
    while (true) {
        cout << endl << "Доступные команды:" << endl;
        cout << "INSERT <ключ> <значение> - добавить элемент" << endl;
        cout << "REMOVE <ключ> - удалить элемент" << endl;
        cout << "FIND <ключ> - найти элемент" << endl;
        cout << "EXIT - выход из программы" << endl;
        cout << "Введите команду:" << endl;
        
        string input;
        getline(cin, input);
        
        stringstream ss(input);
        string command;
        ss >> command;
        
        if (command == "EXIT") {
            cout << "Выход из программы" << endl;
            break;
        }
        else if (command == "INSERT") {
            string key;
            string value;
            if (ss >> key >> value) {
                hashTable.insert(key, value);
            } else {
                cout << "Ошибка: неверный формат команды. Используйте: INSERT <ключ> <значение>" << endl;
            }
        }
        else if (command == "REMOVE") {
            string key;
            if (ss >> key) {
                cout << "Попытка удаления элемента с ключом='" << key << "'..." << endl;
                hashTable.remove(key);
            } else {
                cout << "Ошибка: неверный формат команды. Используйте: REMOVE <ключ>" << endl;
            }
        }
        else if (command == "FIND") {
            string key;
            if (ss >> key) {
                hashTable.find(key);
            } else {
                cout << "Ошибка: неверный формат команды. Используйте: FIND <ключ>" << endl;
            }
        }
        else {
            cout << "Ошибка: неизвестная команда '" << command << "'" << endl;
        }
    }
}

File: hashTableFree.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func genHashFree(size int, key string) int {
	result := 1245
	for i := 0; i < len(key); i++ {
		result += i * int(key[i]) % size
	}
	return result % size
}

type HashTableFree struct {
	table []pairF
	size  int
}

type pairF struct {
	first  string
	second string
}

func NewHashTableFree(sz int) *HashTableFree {
	table := make([]pairF, sz)
	for i := 0; i < sz; i++ {
		table[i].first = "_empty_"
	}
	return &HashTableFree{
		table: table,
		size:  sz,
	}
}

func (h *HashTableFree) insert(key, value string) {
	Hash := genHashFree(h.size, key)
	if h.table[Hash].first == key {
		h.table[Hash] = pairF{key, value}
		fmt.Printf("Элемент добавлен: ключ='%s', значение='%s'\n", key, value)
		return
	}
	if h.table[Hash].first == "_empty_" {
		h.table[Hash] = pairF{key, value}
		fmt.Printf("Элемент добавлен: ключ='%s', значение='%s'\n", key, value)
	} else {
		i := 1
		for {
			index := (Hash + i) % h.size
			if h.table[index].first == key {
				h.table[Hash] = pairF{key, value}
				fmt.Printf("Элемент добавлен: ключ='%s', значение='%s'\n", key, value)
				return
			}
			if h.table[index].first == "_empty_" {
				h.table[index] = pairF{key, value}
				fmt.Printf("Элемент добавлен: ключ='%s', значение='%s'\n", key, value)
				return
			}
			i++
			if i > h.size*2 {
				fmt.Println("Свободное место отсутствует в таблице")
				return
			}
		}
	}
}

func (h *HashTableFree) remove(key string) {
	Hash := genHashFree(h.size, key)
	if h.table[Hash].first == key {
		h.table[Hash].first = "_empty_"
		fmt.Println("Элемент успешно удалён")
	} else {
		i := 1
		for {
			index := (Hash + i) % h.size
			if h.table[index].first == key {
				h.table[index].first = "_empty_"
				fmt.Println("Элемент успешно удалён")
				return
			}
			i++
			if i > h.size*2 {
				fmt.Println("Элемент отсутствует в таблице")
				return
			}
		}
	}
}

func (h *HashTableFree) find(key string) {
	Hash := genHashFree(h.size, key)
	if h.table[Hash].first == key {
		fmt.Printf("Data = %s\n", h.table[Hash].second)
	} else {
		i := 1
		for {
			index := (Hash + i) % h.size
			if h.table[index].first == key {
				fmt.Printf("Data = %s\n", h.table[index].second)
				return
			}
			i++
			if i > h.size*2 {
				fmt.Println("Элемент отсутствует в таблице")
				return
			}
		}
	}
}

func hashTableFree() {
	fmt.Println("Введите размер хеш-таблицы:")

	var tableSize int
	fmt.Scan(&tableSize)

	if tableSize <= 0 {
		fmt.Println("Ошибка: размер таблицы должен быть положительным числом")
		return
	}

	hashTable := NewHashTableFree(tableSize)
	fmt.Printf("Хеш-таблица создана с размером %d\n", tableSize)

	scanner := bufio.NewScanner(os.Stdin)

	for {
		fmt.Println("\nДоступные команды:")
		fmt.Println("INSERT <ключ> <значение> - добавить элемент")
		fmt.Println("REMOVE <ключ> - удалить элемент")
		fmt.Println("FIND <ключ> - найти элемент")
		fmt.Println("EXIT - выход из программы")
		fmt.Println("Введите команду:")

		scanner.Scan()
		input := scanner.Text()

		parts := strings.Fields(input)
		if len(parts) == 0 {
			continue
		}

		command := parts[0]

		if command == "EXIT" {
			fmt.Println("Выход из программы")
			break
		} else if command == "INSERT" {
			if len(parts) >= 3 {
				key := parts[1]
				value := strings.Join(parts[2:], " ")
				hashTable.insert(key, value)
			} else {
				fmt.Println("Ошибка: неверный формат команды. Используйте: INSERT <ключ> <значение>")
			}
		} else if command == "REMOVE" {
			if len(parts) >= 2 {
				key := parts[1]
				fmt.Printf("Попытка удаления элемента с ключом='%s'...\n", key)
				hashTable.remove(key)
			} else {
				fmt.Println("Ошибка: неверный формат команды. Используйте: REMOVE <ключ>")
			}
		} else if command == "FIND" {
			if len(parts) >= 2 {
				key := parts[1]
				hashTable.find(key)
			} else {
				fmt.Println("Ошибка: неверный формат команды. Используйте: FIND <ключ>")
			}
		} else {
			fmt.Printf("Ошибка: неизвестная команда '%s'\n", command)
		}
	}
}
